### PR TITLE
Add Command unlockMap

### PR DESCRIPTION
**Pull request by chengecu**
Use command "unlockMap uid" to unlock all maps and points.
I am not very clear about the openstateid for unlocking the map limit, so I have roughly filled in some. If someone know, please help me improve it.

### CODE REVIEW

File: src/GameServer/Commands/OnUnlockMap.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+namespace Weedwacker.GameServer.Commands
+{
+    public static partial class ConsoleCommands
+    {
+        
+        public static async Task unlockMap(IConsole console,int guid) // GameUid
+        {
+            Player targetPlayer = GameServer.OnlinePlayers[guid].Player;
+            List<int> sceneid = new List<int>();
+            Dictionary<int, ScenePointData> pointdataperscene = new Dictionary<int, ScenePointData>();
```
Comment (akatatsu27) [inline src/GameServer/Commands/OnUnlockMap.cs:22]: your variable names are all over the place. c# local variables start with lowercase, andThenUppercaseAtTheBeginningOfEachWord

File: src/GameServer/Systems/Player/ProgressManager.cs [hunk 2]
```diff
@@ -42,10 +45,71 @@ public async Task<bool> UnlockTransPoint(int sceneId, int pointId)
             var update = Builders<ProgressManager>.Update.Set(w => w.UnlockedScenePoints, UnlockedScenePoints);
             await DatabaseManager.UpdateProgressAsync(filter, update);
             //await Owner.Inventory.AddItemByIdAsync(201, 5, ActionReason.UnlockPointReward);
-            // await Owner.Inventory.AddItemByIdAsync(102, isStatue ? 50 : 10, ActionReason.UnlockPointReward);
+            //await Owner.Inventory.AddItemByIdAsync(102, isStatue ? 50 : 10, ActionReason.UnlockPointReward);
             await Owner.SendPacketAsync(new PacketScenePointUnlockNotify((uint)sceneId, (uint)pointId));
             return true;
         }
+        public async Task<bool> UnlockTransPoint(List<int> sceneId, Dictionary<int, HashSet<uint>> pointIds)
```
Comment (akatatsu27) [inline src/GameServer/Systems/Player/ProgressManager.cs:52]: This function signature is wasteful and error prone
```suggestion
        public async Task<bool> UnlockTransPoint(IDictionary<int, HashSet<uint>> pointIds)
```

File: src/GameServer/Packet/Send/PacketScenePointUnlockNotify.cs [hunk 3]
```diff
@@ -7,7 +7,19 @@ namespace Weedwacker.GameServer.Packet.Send
 {
     internal class PacketScenePointUnlockNotify : BasePacket
     {
-        public PacketScenePointUnlockNotify(uint sceneId,uint pointId) : base(Enums.OpCode.ScenePointUnlockNotify)
+
+
+        public PacketScenePointUnlockNotify(int sceneId, List<uint> pointIds) : base(Enums.OpCode.ScenePointUnlockNotify)
```
Comment (akatatsu27) [inline src/GameServer/Packet/Send/PacketScenePointUnlockNotify.cs:12]: better
```suggestion
        public PacketScenePointUnlockNotify(int sceneId, IEnumerable<uint> pointIds) : base(Enums.OpCode.ScenePointUnlockNotify)
```

File: src/GameServer/Packet/Send/PacketSceneAreaUnlockNotify.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
+﻿using Google.Protobuf;
+using Weedwacker.GameServer.Enums;
+using Weedwacker.Shared.Network.Proto;
+
+namespace Weedwacker.GameServer.Packet.Send
+{
+    internal class PacketSceneAreaUnlockNotify : BasePacket
+    {
+        public PacketSceneAreaUnlockNotify(int sceneId, List<uint> areaIds) : base(Enums.OpCode.SceneAreaUnlockNotify)
```
Comment (akatatsu27) [inline src/GameServer/Packet/Send/PacketSceneAreaUnlockNotify.cs:9]: better
```suggestion
        public PacketSceneAreaUnlockNotify(int sceneId, IEnumerable<uint> areaIds) : base(Enums.OpCode.SceneAreaUnlockNotify)
```

File: src/GameServer/Systems/Player/ProgressManager.cs [hunk 5]
```diff
@@ -42,10 +45,71 @@ public async Task<bool> UnlockTransPoint(int sceneId, int pointId)
...
+                    
+                    if (pointIds[a] != null)
+                    {
+                        UnlockedScenePoints.Add(a, new HashSet<int>());
+                        foreach (int i in pointIds[a])
+                        {
+                            UnlockedScenePoints[a].Add(i);
+                            List<uint> list = new List<uint>();
+                            list.AddRange(pointIds[a]);
+                            await Owner.SendPacketAsync(new PacketScenePointUnlockNotify(a, list));
```
Comment (akatatsu27) [inline src/GameServer/Systems/Player/ProgressManager.cs:67]: with the above changes you can pass pointIds[a] directly
```suggestion
                            await Owner.SendPacketAsync(new PacketScenePointUnlockNotify(a, pointIds[a]));
```

File: src/GameServer/Systems/Player/ProgressManager.cs [hunk 6]
```diff
@@ -42,10 +45,71 @@ public async Task<bool> UnlockTransPoint(int sceneId, int pointId)
...
+                {
+                    
+                    if (pointIds[a] != null)
+                    {
+                        UnlockedScenePoints.Add(a, new HashSet<int>());
+                        foreach (int i in pointIds[a])
+                        {
+                            UnlockedScenePoints[a].Add(i);
+                            List<uint> list = new List<uint>();
+                            list.AddRange(pointIds[a]);
```
Comment (akatatsu27) [inline src/GameServer/Systems/Player/ProgressManager.cs:66]: each class that implements the IEnumerable interface has the ToList<T>() generic function available

File: src/GameServer/Systems/Player/ProgressManager.cs [hunk 7]
```diff
@@ -42,10 +45,71 @@ public async Task<bool> UnlockTransPoint(int sceneId, int pointId)
...
+                    
+                }
+                                   
+            }
+            var filter = Builders<ProgressManager>.Filter.Where(p => p.OwnerId == Owner.GameUid);
+            var update = Builders<ProgressManager>.Update.Set(w => w.UnlockedScenePoints, UnlockedScenePoints);
+            await DatabaseManager.UpdateProgressAsync(filter, update);
+            return true;
+        }
+        public async Task<bool> UnlockedSceneArea(int sceneId, int areaId)
```
Comment (akatatsu27) [inline src/GameServer/Systems/Player/ProgressManager.cs:80]:
```suggestion
        public async Task<bool> UnlockSceneArea(int sceneId, int areaId)
```

File: src/GameServer/Systems/Player/ProgressManager.cs [hunk 8]
```diff
@@ -42,10 +45,71 @@ public async Task<bool> UnlockTransPoint(int sceneId, int pointId)
...
+
+            UnlockedSceneAreas[sceneId].Add(areaId);
+            var filter = Builders<ProgressManager>.Filter.Where(p => p.OwnerId == Owner.GameUid);
+            var update = Builders<ProgressManager>.Update.Set(w => w.UnlockedSceneAreas, UnlockedSceneAreas);
+            await DatabaseManager.UpdateProgressAsync(filter, update);
+            await Owner.SendPacketAsync(new PacketSceneAreaUnlockNotify(sceneId,areaId));
+            return true;
+        }
+
+        public async Task<bool> UnlockedSceneArea(List<int> sceneId, List<uint> areaIds)
```
Comment (akatatsu27) [inline src/GameServer/Systems/Player/ProgressManager.cs:93]: better
```suggestion
        public async Task<bool> UnlockSceneArea(List<int> sceneId, IEnumerable<uint> areaIds)
```

**Review by akatatsu27 — CHANGES_REQUESTED** [review]

**Comment by chengecu** [comment]
Thank you very much for your correction. I will try to improve it

**Pull request closed without merging**